addons/account_edi/models/account_edi_format.py

```diff
@@ -499,21 +499,91 @@ def _retrieve_partner(self, name=None, phone=None, mail=None, vat=None):
         :param vat:     The vat number of the partner.
         :returns:       A partner or an empty recordset if not found.
         '''
-        domains = []
-        for value, domain in (
-            (name, [('name', 'ilike', name)]),
-            (phone, expression.OR([[('phone', '=', phone)], [('mobile', '=', phone)]])),
-            (mail, [('email', '=', mail)]),
-            (vat, [('vat', 'like', vat)]),
-        ):
-            if value is not None:
-                domains.append(domain)
-
-        domain = expression.AND([
-            expression.OR(domains),
-            [('company_id', 'in', [False, self.env.company.id])],
-        ])
-        return self.env['res.partner'].search(domain, limit=1)
+        def search_with_vat(extra_domain):
+            if not vat:
+                return None
+
+            # Sometimes, the vat is specified with some whitespaces.
+            normalized_vat = vat.replace(' ', '')
+
+            partner = self.env['res.partner'].search([('vat', '=', normalized_vat)], limit=1)
+
+            # Try to remove the country code prefix from the vat.
+            if not partner and len(normalized_vat) > 2 and normalized_vat[:2].isalpha():
+                country_prefix = normalized_vat[:2]
+                normalized_vat = normalized_vat[2:]
+                partner = self.env['res.partner'].search(extra_domain + [
+                    ('vat', '=', normalized_vat),
+                    ('country_id', '!=', False),
+                    ('country_id.code', '=', country_prefix.lower()),
+                ], limit=1)
+
+                # The country could be not specified.
+                if not partner:
+                    partner = self.env['res.partner'].search(extra_domain + [
+                        ('vat', '=', normalized_vat),
+                        ('country_id', '=', False),
+                    ], limit=1)
+
+            # The vat could be a string of alphanumeric values without country code but with missing zeros at the
+            # beginning.
+            if not partner:
+                try:
+                    vat_only_alphanumeric = str(int(normalized_vat))
+
+                    query = self.env['res.partner']._where_calc([('active', '=', True)], extra_domain)
+                    tables, where_clause, where_params = query.get_sql()
+
+                    self._cr.execute(f'''
+                        SELECT res_partner.id
+                        FROM {tables}
+                        WHERE {where_clause}
+                        AND res_partner.vat ~ %s
+                        LIMIT 1
+                    ''', where_params + ['^0*%s$' % vat_only_alphanumeric])
```

Comment:
This thing is notoriously too complicated.    What is wrong with '|', '|', ('vat', '=', vat0), ('vat', '=', vat1) 

In most countries, vat number is the most important one anyways and even searching for phone is weird.  Also, it should be sufficient to do company_id in (False, self.env.company_id)  

But it is good to have a generic method everyone can use, but the complexity seems so exaggerated.  (of course there might be reasons, but even) @smetl